File: old/1 - estruturaDados/3 - trab octree/oct.py
import numpy as np

class OctreeNode:
    def __init__(self, center, size, depth=0, max_depth=5, max_elements=8):
        """
        Inicializa um nó da Octree.
        
        Args:
            center (tuple): Centro do nó (x, y, z)
            size (float): Tamanho do nó (largura/altura/profundidade)
            depth (int): Profundidade atual do nó
            max_depth (int): Profundidade máxima permitida
            max_elements (int): Número máximo de elementos antes de subdividir
        """
        self.center = np.array(center)
        self.size = size
        self.depth = depth
        self.max_depth = max_depth
        self.max_elements = max_elements
        self.elements = []
        self.children = []
        self.is_divided = False
    
    def subdivide(self):
        """Subdivide o nó em 8 filhos."""
        if self.is_divided or self.depth >= self.max_depth:
            return
        
        child_size = self.size / 2
        offset = child_size / 2
        
        # Cria os 8 octantes
        for x in [-1, 1]:
            for y in [-1, 1]:
                for z in [-1, 1]:
                    child_center = self.center + np.array([x, y, z]) * offset
                    self.children.append(
                        OctreeNode(
                            child_center, 
                            child_size, 
                            self.depth + 1, 
                            self.max_depth, 
                            self.max_elements
                        )
                    )
        
        self.is_divided = True
        
        # Redistribui os elementos para os filhos
        for element in self.elements:
            for child in self.children:
                if child.contains(element['point']):
                    child.insert(element['point'], element['data'])
                    break
        
        self.elements = []
    
    def contains(self, point):
        """Verifica se um ponto está dentro deste nó."""
        point = np.array(point)
        return np.all(np.abs(point - self.center) <= self.size/2)
    
    def find_child(self, point):
        """Encontra o filho que contém o ponto especificado."""
        for child in self.children:
            if child.contains(point):
                return child
        return None
    
    def insert(self, point, data):
        """Insere um ponto com dados associados na Octree."""
        point = np.array(point)
        
        if not self.contains(point):
            return False
        
        if not self.is_divided and (len(self.elements) < self.max_elements or self.depth == self.max_depth):
            self.elements.append({'point': point, 'data': data})
            return True
        
        if not self.is_divided:
            self.subdivide()
        
        child = self.find_child(point)
        if child:
            return child.insert(point, data)
        
        return False
    
    def query_range(self, center, size):
        """Consulta todos os elementos dentro de um range especificado."""
        results = []
        if not self.intersects(center, size):
            return results
        
        if self.is_divided:
            for child in self.children:
                results.extend(child.query_range(center, size))
        else:
            for element in self.elements:
                point = element['point']
                if np.all(np.abs(point - center) <= size/2):
                    results.append(element)
        
        return results
    
    def intersects(self, center, size):
        """Verifica se este nó intersecta com um range especificado."""
        center = np.array(center)
        return np.all(np.abs(self.center - center) <= (self.size + size)/2)
    
    def __str__(self):
        """Representação em string do nó."""
        indent = '  ' * self.depth
        s = f"{indent}Node(depth={self.depth}, center={self.center}, size={self.size}, elements={len(self.elements)}"
        if self.is_divided:
            for child in self.children:
                s += '\n' + str(child)
        return s


class Octree:
    def __init__(self, center, size, max_depth=5, max_elements=8):
        """
        Inicializa a Octree.
        
        Args:
            center (tuple): Centro da Octree (x, y, z)
            size (float): Tamanho inicial da Octree
            max_depth (int): Profundidade máxima
            max_elements (int): Número máximo de elementos por nó antes de subdividir
        """
        self.root = OctreeNode(center, size, max_depth=max_depth, max_elements=max_elements)
    
    def insert(self, point, data=None):
        """Insere um ponto na Octree."""
        return self.root.insert(point, data)
    
    def query_range(self, center, size):
        """Consulta pontos dentro de um range especificado."""
        return self.root.query_range(center, size)
    
    def __str__(self):
        """Representação em string da Octree."""
        return str(self.root)


# Exemplo de uso
if __name__ == "__main__":
    # Cria uma Octree centrada na origem com tamanho 20
    octree = Octree(center=(0, 0, 0), size=20, max_depth=4)
    
    # Insere alguns pontos aleatórios
    np.random.seed(42)
    for i in range(100):
        point = np.random.uniform(-10, 10, size=3)
        octree.insert(point, f"objeto_{i}")
    
    # Consulta pontos em um range
    print("Pontos no range [-5,-5,-5] a [5,5,5]:")
    results = octree.query_range(center=(0, 0, 0), size=10)
    print(f"Encontrados {len(results)} pontos")
    
    # Imprime a estrutura da Octree (para debug)
    # print(octree)

File: old/1 - estruturaDados/3 - trab octree/em c#/oct.cs
using UnityEngine;
using System.Collections.Generic;

public class OctreeNode
{
    // Limite de objetos antes de subdividir
    public int Capacity = 4; // se atingir o limite, se subdivide 
    // Profundidade máxima da árvore
    public int MaxDepth = 5; // limite pra evitar subdivisão infinita 

    // Bounding Box (limites do nó)
    public Bounds Bounds { get; private set; } // delimita 3d (X; Y; Z;) || (0,0,0) a (10,10,10)
    // Objetos armazenados neste nó (se for uma folha)
    public List<Vector3> Objects { get; private set; } // os pontos armazenados nesse nó 
    // Filhos (8 subdivisões)
    public OctreeNode[] Children { get; private set; } // subdivs
    // Profundidade atual
    public int Depth { get; private set; } //profundidade atual  

    public OctreeNode(Bounds bounds, int depth) // CONSTRUTOR 
    {
        Bounds = bounds; // espaço 3d q ele cobre 
        Depth = depth; // o nivel de profundidade, sendo 0 a raiz 
        Objects = new List<Vector3>(); // lista vazia pra armazenar pontos  
        Children = null; // começa null pq ainda nao foi subdividido, ou seja nao tem filho
    }

    // Insere um ponto na Octree, aqui q começa a feder 
    public bool Insert(Vector3 point)
    {
        // Se o ponto não estiver dentro deste nó, ignore
        // traduzindo: verifica se o ponto está dentro dos limites do nó 
        if (!Bounds.Contains(point))
            return false;

        // Se há espaço ou ainda não atingiu a capacidade máxima, adicione ao nó atual
        // traduzindo: self-explanatory actually 
        if (Objects.Count < Capacity || Depth >= MaxDepth)
        {
            Objects.Add(point);
            return true;
        }

        // Se não tem filhos ainda, subdivida
        // traduzindo: self-explanatory actually 
        if (Children == null)
            Subdivide(); // chama pra criar 8 filhos 

        // Tente inserir em um dos filhos
        // traduzindo: tenta inserir usando recursão 
        foreach (var child in Children)
        {
            if (child.Insert(point))
                return true;
        }

        return false;
    }

    // Subdivide o nó em 8 filhos
    private void Subdivide()
    {
        Children = new OctreeNode[8];
        Vector3 size = Bounds.size / 2;
        Vector3 center = Bounds.center;

        // cria 8 filhos, cobrindo todos os octantes 
        for (int i = 0; i < 8; i++)
        {
            Vector3 newCenter = center; // calcula o centro e tamanho de cada filho, usando operações bit-a-bit pra determinar a posição de cada octante, b-a-b = algebra booleana, usando And Or Xor e etc, aqueles calculo legal la
            newCenter.x += (i & 1) == 0 ? -size.x / 2 : size.x / 2;
            newCenter.y += (i & 2) == 0 ? -size.y / 2 : size.y / 2;
            newCenter.z += (i & 4) == 0 ? -size.z / 2 : size.z / 2;

            Bounds childBounds = new Bounds(newCenter, size);
            Children[i] = new OctreeNode(childBounds, Depth + 1);
        }

        // Redistribui os objetos antigos para os filhos, cada obj
        foreach (var obj in Objects)
        {
            foreach (var child in Children)
            {
                if (child.Insert(obj))
                    break;
            }
        }
        Objects.Clear(); // limpa os objetos do nó atual 
    }

    // Desenha os limites da Octree (para debug no Unity)
    // cria a visualização pra unity 
    public void DrawGizmos() 
    {
        Gizmos.color = Color.green;
        Gizmos.DrawWireCube(Bounds.center, Bounds.size);

        if (Children != null)
        {
            foreach (var child in Children)
            {
                child.DrawGizmos();
            }
        }
    }
}

// Demo pratica na unity 
public class OctreeDemo : MonoBehaviour
{
    public int MaxPoints = 100; // numero de pontos a gerar 
    public float WorldSize = 10f; // tamanho do mundo 3D
    private OctreeNode octree; // Raiz da Octree 
    private List<Vector3> points = new List<Vector3>(); // Pontos gerados 

    void Start()
    {
        // inicia a octree com um bounding box cobrindo o mundo (caixa delimitadora, traduzindo), em worldSize 
        Bounds worldBounds = new Bounds(Vector3.zero, Vector3.one * WorldSize);
        octree = new OctreeNode(worldBounds, 0);

        // Gera pontos aleatórios e insere na Octree
        for (int i = 0; i < MaxPoints; i++)
        {
            Vector3 randomPoint = new Vector3(
                Random.Range(-WorldSize / 2, WorldSize / 2),
                Random.Range(-WorldSize / 2, WorldSize / 2),
                Random.Range(-WorldSize / 2, WorldSize / 2)
            );
            points.Add(randomPoint);
            octree.Insert(randomPoint);
        }
    }

    void OnDrawGizmos()
    {
        if (octree != null)
        {
            octree.DrawGizmos(); // desenha a octree 
            Gizmos.color = Color.red;
            foreach (var point in points)
            {
                Gizmos.DrawSphere(point, 0.1f); // desenha os pontos
            }
        }
    }
}